test/utils/diffpatcher.spec.js

```diff
@@ -0,0 +1,94 @@
+import test from 'tape'
+import * as diffpatcher from '../../lib/sync/utils/diffpatcher'
+
+test('Utils::diffpatcher', t => {
+  t.test('::diff', t => {
+    t.test('should diff a single object', t => {
+      const original = { name: 'old' }
+      const update = { name: 'new' }
+      const diff = diffpatcher.diff(original, update)
+      t.deepEqual(diff, { name: ['old', 'new'] })
+      t.end()
+    })
+
+    t.test('should diff a list of scalar values', t => {
+      const original = ['old', 'list']
+      const update = ['new', 'list']
+      const diff = diffpatcher.diff(original, update)
+      // expect first value of the list to be modified
+      const delta = diff[0]
+      t.deepEqual(delta, ['new'])
+      t.end()
+    })
+
+    t.test('should provide correct index on the array', t => {
+      const original = [
+        { name: 'old' },
+        { name: 'list' },
+      ]
+      const update = [
+        null,
+        { name: 'list' },
+      ]
+
+      const diff = diffpatcher.diff(original, update)
+      // expect first value of the list to be modified
+      const delta = diff[0]
+      t.deepEqual(delta, [ null ])
+      t.end()
+    })
+    t.end()
+  })
+
+  t.test('::objectHash', t => {
+    t.test(
+      'should provide index value if obj is not defined or null in a list',
+      t => {
+        const objectHash = diffpatcher.objectHash
+        const randomValues = [
+          { name: 'random1' },
+          null,
+          '',
+          0,
+          [],
+        ]
+
+        const objectHashIndexValues = randomValues
+          .map((r, i) => objectHash(r, i))
+
+        t.deepEqual(objectHashIndexValues, [
+          'random1',
+          '$$index:1',
+          '$$index:2',
+          '$$index:3',
+          '$$index:4',
+        ], 'expect index value OR hash of the object in the list')
+        t.end()
```

Comment:
Sorry, this is not exactly what I meant. A test with product data that reproduces the error, using buildActions main function. See sync tests, or ask @PhilippSpo